src/app/tests/TestClusterObjectUtils.cpp

```diff
@@ -0,0 +1,853 @@
+static SecureSessionMgr gSessionManager;
+static Messaging::ExchangeManager gExchangeManager;
+static secure_channel::MessageCounterManager gMessageCounterManager;
+static TransportMgr<Transport::UDP> gTransportManager;
+
+namespace app {
+
+class TestClusterObjectUtils
+{
+public:
+    static void TestClusterObjectUtils_EncAndDecSimpleStruct(nlTestSuite * apSuite, void * apContext);
+    static void TestClusterObjectUtils_EncAndDecNestedStruct(nlTestSuite * apSuite, void * apContext);
+    static void TestClusterObjectUtils_EncAndDecNestedStructList(nlTestSuite * apSuite, void * apContext);
+    static void TestClusterObjectUtils_EncAndDecIterableNestedStructList(nlTestSuite * apSuite, void * apContext);
+    static void TestClusterObjectUtils_EncAndDecIterableDoubleNestedStructList(nlTestSuite * apSuite, void * apContext);
+
+    static void TestClusterObjectUtils_OptionalFields(nlTestSuite * apSuite, void * apContext);
+    static void TestClusterObjectUtils_ExtraField(nlTestSuite * apSuite, void * apContext);
+    static void TestClusterObjectUtils_InvalidSimpleFieldTypes(nlTestSuite * apSuite, void * apContext);
+    static void TestClusterObjectUtils_InvalidListType(nlTestSuite * apSuite, void * apContext);
+
+private:
+    void SetupBuf();
+    void DumpBuf();
+    void SetupReader();
+
+    chip::System::TLVPacketBufferBackingStore mStore;
+    chip::TLV::TLVWriter mWriter;
+    chip::TLV::TLVReader mReader;
+    nlTestSuite * mpSuite;
+};
+
+using namespace chip::TLV;
+
+TestClusterObjectUtils gTestClusterObjectUtils;
+
+void TestClusterObjectUtils::SetupBuf()
+{
+    chip::System::PacketBufferHandle buf;
+
+    buf = System::PacketBufferHandle::New(1024);
+    mStore.Init(std::move(buf));
+
+    mWriter.Init(mStore);
+    mReader.Init(mStore);
+}
+
+void TestClusterObjectUtils::DumpBuf()
+{
+    TLV::TLVReader reader;
+    reader.Init(mStore);
+
+    //
+    // TODO: Check-in TLV Text pretty printer in separate PR
+    //
+    // chip::TLV::Utilities::Print(reader);
+}
+
+void TestClusterObjectUtils::SetupReader()
+{
+    CHIP_ERROR err;
+
+    mReader.Init(mStore);
+    err = mReader.Next();
+
+    NL_TEST_ASSERT(mpSuite, err == CHIP_NO_ERROR);
+}
+
+template <typename T>
+struct TagValuePair
+{
+    uint64_t tag;
+    T & value;
+};
+
+template <typename T>
+TagValuePair<T> MakeTagValuePair(uint64_t tag, T & value)
+{
+    return TagValuePair<T>{ tag, value };
+}
+
+template <typename... ArgTypes>
+CHIP_ERROR EncodeStruct(TLV::TLVWriter & writer, uint64_t tag, ArgTypes... Args)
+{
+    using expand_type = int[];
+    TLV::TLVType type;
+
+    ReturnErrorOnFailure(writer.StartContainer(tag, TLV::kTLVType_Structure, type));
+    (void) (expand_type{ (DataModel::Encode(writer, Args.tag, Args.value), 0)... });
+    ReturnErrorOnFailure(writer.EndContainer(type));
+
+    return CHIP_NO_ERROR;
+}
+
+bool StringMatches(Span<const char> str1, const char * str2)
+{
+    if ((str1.data() == nullptr) ^ (str2 == nullptr))
```

Comment:
Check one verifies that they are different, check 2 verifies that none are null... so is it true that only 2nd test needs to be done? If 2nd test fails, none are null, so this comparison compares false ^ false